File: src/agentegrity/core/topology.py
"""
Agent topology - the in-process multi-agent system snapshot.

Agentegrity models multi-agent systems as immutable topologies.
Each :class:`AgentTopology` is a hashable snapshot of which agents
participate, what roles they hold, and which communication channels
exist between them. Mutation produces a new snapshot with a new
content hash; the old snapshot is unchanged.

Immutability is load-bearing for two invariants:

1. Recovery integrity needs deterministic restore targets. A mutable
   topology means "restore to checkpoint X" is undefined when the
   topology since changed. Snapshot-and-replace makes the restore
   target an explicit object.
2. Attestation records commit to *which* topology was live at
   evaluation time. The attestation carries a reference (Evidence of
   type ``"topology"``) whose content hash pins the snapshot.

The :class:`TopologyKind` enum names the four shapes the supported
adapters expose: hub-spoke (CrewAI sequential, Agno coordinator,
Bedrock supervisor), hierarchical DAG (LangGraph workflows, Google
ADK), peer-to-peer (LangGraph swarm, OpenAI Swarm, Agno
collaborator), and group chat (AutoGen). Adapters declare their
topology kind at instrument time and the layers consume it through
``_ContextBuffer.to_evaluation_context()``.

A dynamic team (e.g. Agno's callable members provider) is
represented as a sequence of topology snapshots; each membership
change emits a ``topology_change`` event and a fresh
:class:`AgentTopology` lands in the chain.
"""

from __future__ import annotations

import hashlib
import json
import uuid
from dataclasses import dataclass, field, replace
from datetime import datetime, timezone
from enum import Enum
from typing import Any


class AgentRole(str, Enum):
    """The role an agent holds within its topology.

    Naming is deliberately framework-agnostic. Hub-spoke topologies
    use ``LEADER`` for the orchestrator and ``MEMBER`` for the
    workers. Hierarchical DAGs use ``SUPERVISOR`` and ``WORKER``.
    Peer-to-peer topologies use ``PEER`` throughout.

    The role drives the Cortical layer's role-conformance check
    (the same agent in different roles gets different behavioural
    baselines) and the Federation layer's topology-coherence score
    (LEADER tool-call frequency vs WORKER tool-call frequency).
    """

    LEADER = "leader"
    MEMBER = "member"
    SUPERVISOR = "supervisor"
    WORKER = "worker"
    PEER = "peer"


class TopologyKind(str, Enum):
    """The shape of a multi-agent topology.

    Maps onto the framework landscape:

    * ``HUB_SPOKE``: one orchestrator + N workers. CrewAI sequential
      crews, Agno teams in coordinator mode, AWS Bedrock supervisor
      with collaborators.
    * ``HIERARCHICAL_DAG``: arbitrary parent-child tree.
      LangGraph workflow graphs, Google ADK ``SequentialAgent`` /
      ``ParallelAgent`` / ``LoopAgent`` compositions.
    * ``PEER_TO_PEER``: no orchestrator, handoffs between equals.
      LangGraph swarm, OpenAI Swarm, Agno collaborator mode.
    * ``GROUP_CHAT``: dynamic membership with broadcast comms.
      AutoGen ``GroupChatManager``-driven conversations.
    """

    HUB_SPOKE = "hub_spoke"
    HIERARCHICAL_DAG = "hierarchical_dag"
    PEER_TO_PEER = "peer_to_peer"
    GROUP_CHAT = "group_chat"


@dataclass(frozen=True)
class AgentMember:
    """One agent's place in a topology snapshot.

    ``capabilities`` is a tuple (not a list) so the dataclass is
    hashable. ``parent_id`` is the immediate parent's ``agent_id``
    or ``None`` for the topology's root (the leader in HUB_SPOKE,
    the supervisor of a HIERARCHICAL_DAG, etc.). In PEER_TO_PEER
    and GROUP_CHAT topologies every member has ``parent_id=None``.
    """

    agent_id: str
    name: str
    role: AgentRole
    parent_id: str | None = None
    capabilities: tuple[str, ...] = ()

    def to_dict(self) -> dict[str, Any]:
        return {
            "agent_id": self.agent_id,
            "name": self.name,
            "role": self.role.value,
            "parent_id": self.parent_id,
            "capabilities": list(self.capabilities),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "AgentMember":
        return cls(
            agent_id=data["agent_id"],
            name=data["name"],
            role=AgentRole(data["role"]),
            parent_id=data.get("parent_id"),
            capabilities=tuple(data.get("capabilities", ())),
        )


@dataclass(frozen=True)
class AgentTopology:
    """An immutable snapshot of a multi-agent topology.

    Constructed once at instrument time; mutations produce a new
    snapshot via :meth:`with_member` / :meth:`without_member`.
    ``content_hash()`` returns a deterministic SHA-256 over the
    canonical JSON; this hash is what attestations commit to via
    ``Evidence(evidence_type="topology", ...)``.

    ``topology_id`` is generated at construction and stays stable
    across mutations of the same logical topology, so a verifier
    can correlate a sequence of snapshots. ``content_hash`` differs
    on every mutation.
    """

    kind: TopologyKind
    members: tuple[AgentMember, ...] = ()
    comm_channels: frozenset[str] = field(default_factory=frozenset)
    topology_id: str = field(default_factory=lambda: str(uuid.uuid4()))
    created_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def leader(self) -> AgentMember | None:
        """The orchestrator member, if one exists.

        Hub-spoke and hierarchical-DAG topologies have one;
        peer-to-peer and group-chat return ``None``.
        """
        for member in self.members:
            if member.role in (AgentRole.LEADER, AgentRole.SUPERVISOR):
                return member
        return None

    def children_of(self, agent_id: str) -> tuple[AgentMember, ...]:
        """Members whose ``parent_id`` equals ``agent_id``."""
        return tuple(m for m in self.members if m.parent_id == agent_id)

    def member(self, agent_id: str) -> AgentMember | None:
        """Look up a member by ``agent_id``."""
        for m in self.members:
            if m.agent_id == agent_id:
                return m
        return None

    @property
    def canonical_payload(self) -> str:
        """Deterministic JSON serialization used for content hashing.

        ``topology_id`` and ``created_at`` are NOT in the payload —
        the hash captures the structural shape (kind, members,
        channels), so adding the same member twice in two different
        sessions produces the same hash. Use ``topology_id`` to
        correlate the lineage; use ``content_hash`` to identify the
        structure.
        """
        payload = {
            "kind": self.kind.value,
            "members": [m.to_dict() for m in self.members],
            "comm_channels": sorted(self.comm_channels),
        }
        return json.dumps(payload, sort_keys=True, separators=(",", ":"))

    def content_hash(self) -> str:
        """SHA-256 of the canonical payload."""
        return hashlib.sha256(self.canonical_payload.encode()).hexdigest()

    def with_member(self, member: AgentMember) -> "AgentTopology":
        """Return a new snapshot with ``member`` added.

        Preserves ``topology_id`` (same logical topology) but
        produces a fresh ``content_hash``. If a member with the
        same ``agent_id`` already exists, it is replaced.
        """
        existing = tuple(m for m in self.members if m.agent_id != member.agent_id)
        return replace(self, members=existing + (member,))

    def without_member(self, agent_id: str) -> "AgentTopology":
        """Return a new snapshot with the member removed."""
        return replace(
            self,
            members=tuple(m for m in self.members if m.agent_id != agent_id),
        )

    def with_channels(self, *channels: str) -> "AgentTopology":
        """Return a new snapshot with additional comm channels."""
        return replace(self, comm_channels=self.comm_channels | frozenset(channels))

    def to_dict(self) -> dict[str, Any]:
        return {
            "topology_id": self.topology_id,
            "kind": self.kind.value,
            "members": [m.to_dict() for m in self.members],
            "comm_channels": sorted(self.comm_channels),
            "created_at": self.created_at.isoformat(),
            "content_hash": self.content_hash(),
        }

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "AgentTopology":
        """Rebuild a topology from its ``to_dict`` representation.

        ``content_hash`` in the input is ignored — it's derived from
        the canonical payload on demand. ``created_at`` is parsed
        from ISO format; if absent (older serializations), defaults
        to the current time.
        """
        created_at = (
            datetime.fromisoformat(data["created_at"])
            if "created_at" in data
            else datetime.now(timezone.utc)
        )
        return cls(
            topology_id=data.get("topology_id", str(uuid.uuid4())),
            kind=TopologyKind(data["kind"]),
            members=tuple(AgentMember.from_dict(m) for m in data.get("members", [])),
            comm_channels=frozenset(data.get("comm_channels", [])),
            created_at=created_at,
        )

    def __repr__(self) -> str:
        return (
            f"AgentTopology(kind={self.kind.value}, "
            f"members={len(self.members)}, "
            f"channels={len(self.comm_channels)}, "
            f"id={self.topology_id[:8]}...)"
        )


@dataclass(frozen=True)
class TopologyChange:
    """A structural diff between two topology snapshots.

    Emitted by adapters when the topology mutates at runtime (e.g.
    Agno's callable members provider produces a new member; AutoGen
    sees a new nested span). The change is recorded in the chain as
    ``Evidence(evidence_type="topology_change", ...)`` so a verifier
    can walk the history of structural shifts.
    """

    previous_topology_id: str
    previous_content_hash: str
    new_topology_id: str
    new_content_hash: str
    added_members: tuple[AgentMember, ...] = ()
    removed_member_ids: tuple[str, ...] = ()

    @classmethod
    def between(
        cls, previous: AgentTopology, current: AgentTopology
    ) -> "TopologyChange":
        """Compute the diff from ``previous`` to ``current``."""
        prev_ids = {m.agent_id for m in previous.members}
        curr_ids = {m.agent_id for m in current.members}
        added = tuple(m for m in current.members if m.agent_id not in prev_ids)
        removed = tuple(prev_ids - curr_ids)
        return cls(
            previous_topology_id=previous.topology_id,
            previous_content_hash=previous.content_hash(),
            new_topology_id=current.topology_id,
            new_content_hash=current.content_hash(),
            added_members=added,
            removed_member_ids=removed,
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "previous_topology_id": self.previous_topology_id,
            "previous_content_hash": self.previous_content_hash,
            "new_topology_id": self.new_topology_id,
            "new_content_hash": self.new_content_hash,
            "added_members": [m.to_dict() for m in self.added_members],
            "removed_member_ids": list(self.removed_member_ids),
        }

File: clients/typescript/packages/client/src/topology.ts
/**
 * `AgentTopology` — TypeScript mirror of Python's
 * `agentegrity.core.topology.AgentTopology`. Immutable snapshot of an
 * in-process multi-agent system.
 *
 * Mutations produce new snapshots via `withMember` / `withoutMember` /
 * `withChannels`. `contentHash()` returns a deterministic SHA-256 hex
 * digest of the canonical JSON — matches the Python implementation
 * exactly so a TS-produced topology snapshot Evidence references the
 * same hash a Python verifier would compute over the equivalent
 * topology.
 *
 * The TS Adapter (`DefaultAdapter`) declares a topology via
 * `setTopology(topology, myRole?)`; the first call emits a
 * `topology_declared` event, subsequent calls with a structurally-
 * distinct topology emit `topology_change`.
 */

import { createHash, randomUUID } from "node:crypto";

export enum AgentRole {
  LEADER = "leader",
  MEMBER = "member",
  SUPERVISOR = "supervisor",
  WORKER = "worker",
  PEER = "peer",
}

export enum TopologyKind {
  HUB_SPOKE = "hub_spoke",
  HIERARCHICAL_DAG = "hierarchical_dag",
  PEER_TO_PEER = "peer_to_peer",
  GROUP_CHAT = "group_chat",
}

export interface AgentMemberInit {
  agentId: string;
  name: string;
  role: AgentRole;
  parentId?: string | null;
  capabilities?: readonly string[];
}

/**
 * One agent's place in a topology snapshot. Immutable: constructed
 * once, frozen, never mutated. Use {@link AgentTopology.withMember}
 * to derive a new topology with this member added or replaced.
 */
export class AgentMember {
  readonly agentId: string;
  readonly name: string;
  readonly role: AgentRole;
  readonly parentId: string | null;
  readonly capabilities: readonly string[];

  constructor(init: AgentMemberInit) {
    this.agentId = init.agentId;
    this.name = init.name;
    this.role = init.role;
    this.parentId = init.parentId ?? null;
    this.capabilities = Object.freeze([...(init.capabilities ?? [])]);
    Object.freeze(this);
  }

  toDict(): Record<string, unknown> {
    return {
      agent_id: this.agentId,
      name: this.name,
      role: this.role,
      parent_id: this.parentId,
      capabilities: [...this.capabilities],
    };
  }

  static fromDict(data: Record<string, unknown>): AgentMember {
    return new AgentMember({
      agentId: String(data.agent_id),
      name: String(data.name),
      role: data.role as AgentRole,
      parentId: (data.parent_id as string | null | undefined) ?? null,
      capabilities: (data.capabilities as string[] | undefined) ?? [],
    });
  }
}

export interface AgentTopologyInit {
  kind: TopologyKind;
  members?: readonly AgentMember[];
  commChannels?: ReadonlySet<string> | readonly string[];
  topologyId?: string;
  createdAt?: string;
}

/**
 * Immutable in-process multi-agent topology snapshot. Mutations
 * produce new snapshots; the old snapshot is unchanged.
 *
 * Immutability is load-bearing for two invariants Python's
 * `AgentTopology` documents:
 * 1. Recovery integrity needs deterministic restore targets; a
 *    mutable topology makes "restore to checkpoint X" undefined.
 * 2. Attestations commit to *which* topology was live at
 *    evaluation time via Evidence(evidence_type="topology",
 *    content_hash=topology.contentHash()).
 */
export class AgentTopology {
  readonly kind: TopologyKind;
  readonly members: readonly AgentMember[];
  readonly commChannels: ReadonlySet<string>;
  readonly topologyId: string;
  readonly createdAt: string;

  constructor(init: AgentTopologyInit) {
    this.kind = init.kind;
    this.members = Object.freeze([...(init.members ?? [])]);
    this.commChannels = Object.freeze(
      new Set(
        init.commChannels instanceof Set
          ? [...init.commChannels]
          : [...(init.commChannels ?? [])],
      ),
    );
    this.topologyId = init.topologyId ?? randomUUID();
    this.createdAt = init.createdAt ?? new Date().toISOString();
    Object.freeze(this);
  }

  /** The orchestrator member, if one exists (HUB_SPOKE / HIERARCHICAL_DAG). */
  leader(): AgentMember | null {
    for (const m of this.members) {
      if (m.role === AgentRole.LEADER || m.role === AgentRole.SUPERVISOR) {
        return m;
      }
    }
    return null;
  }

  /** Members whose parentId equals agentId. */
  childrenOf(agentId: string): readonly AgentMember[] {
    return this.members.filter((m) => m.parentId === agentId);
  }

  /** Look up a member by agentId. */
  member(agentId: string): AgentMember | null {
    for (const m of this.members) {
      if (m.agentId === agentId) return m;
    }
    return null;
  }

  /**
   * Deterministic JSON serialization used for content hashing. The
   * topologyId and createdAt are NOT included — the hash captures
   * the structural shape (kind, members, channels) so two sessions
   * producing the same topology structurally produce the same hash.
   * Matches Python `AgentTopology.canonical_payload`.
   */
  canonicalPayload(): string {
    const payload = {
      kind: this.kind,
      members: this.members.map((m) => m.toDict()),
      comm_channels: [...this.commChannels].sort(),
    };
    return canonicalJsonStringify(payload);
  }

  /** SHA-256 hex digest of the canonical payload. */
  contentHash(): string {
    return createHash("sha256")
      .update(this.canonicalPayload())
      .digest("hex");
  }

  /**
   * Return a new snapshot with `member` added. Preserves topologyId
   * (same logical topology) but produces a fresh content_hash. If a
   * member with the same agentId already exists, it is replaced.
   */
  withMember(member: AgentMember): AgentTopology {
    const filtered = this.members.filter(
      (m) => m.agentId !== member.agentId,
    );
    return new AgentTopology({
      kind: this.kind,
      members: [...filtered, member],
      commChannels: this.commChannels,
      topologyId: this.topologyId,
      createdAt: this.createdAt,
    });
  }

  withoutMember(agentId: string): AgentTopology {
    return new AgentTopology({
      kind: this.kind,
      members: this.members.filter((m) => m.agentId !== agentId),
      commChannels: this.commChannels,
      topologyId: this.topologyId,
      createdAt: this.createdAt,
    });
  }

  withChannels(...channels: string[]): AgentTopology {
    const merged = new Set(this.commChannels);
    for (const c of channels) merged.add(c);
    return new AgentTopology({
      kind: this.kind,
      members: this.members,
      commChannels: merged,
      topologyId: this.topologyId,
      createdAt: this.createdAt,
    });
  }

  toDict(): Record<string, unknown> {
    return {
      topology_id: this.topologyId,
      kind: this.kind,
      members: this.members.map((m) => m.toDict()),
      comm_channels: [...this.commChannels].sort(),
      created_at: this.createdAt,
      content_hash: this.contentHash(),
    };
  }

  static fromDict(data: Record<string, unknown>): AgentTopology {
    return new AgentTopology({
      kind: data.kind as TopologyKind,
      members: ((data.members as Record<string, unknown>[]) ?? []).map(
        AgentMember.fromDict,
      ),
      commChannels: new Set(
        (data.comm_channels as string[] | undefined) ?? [],
      ),
      topologyId: (data.topology_id as string | undefined) ?? randomUUID(),
      createdAt:
        (data.created_at as string | undefined) ?? new Date().toISOString(),
    });
  }
}

export interface TopologyChangeData {
  previousTopologyId: string;
  previousContentHash: string;
  newTopologyId: string;
  newContentHash: string;
  addedMembers: readonly AgentMember[];
  removedMemberIds: readonly string[];
}

/**
 * A structural diff between two topology snapshots. Emitted when
 * the topology mutates at runtime. Carried on the chain as
 * Evidence(evidence_type="topology_change").
 */
export class TopologyChange {
  readonly previousTopologyId: string;
  readonly previousContentHash: string;
  readonly newTopologyId: string;
  readonly newContentHash: string;
  readonly addedMembers: readonly AgentMember[];
  readonly removedMemberIds: readonly string[];

  constructor(data: TopologyChangeData) {
    this.previousTopologyId = data.previousTopologyId;
    this.previousContentHash = data.previousContentHash;
    this.newTopologyId = data.newTopologyId;
    this.newContentHash = data.newContentHash;
    this.addedMembers = Object.freeze([...data.addedMembers]);
    this.removedMemberIds = Object.freeze([...data.removedMemberIds]);
    Object.freeze(this);
  }

  static between(
    previous: AgentTopology,
    current: AgentTopology,
  ): TopologyChange {
    const prevIds = new Set(previous.members.map((m) => m.agentId));
    const currIds = new Set(current.members.map((m) => m.agentId));
    const added = current.members.filter((m) => !prevIds.has(m.agentId));
    const removed = [...prevIds].filter((id) => !currIds.has(id));
    return new TopologyChange({
      previousTopologyId: previous.topologyId,
      previousContentHash: previous.contentHash(),
      newTopologyId: current.topologyId,
      newContentHash: current.contentHash(),
      addedMembers: added,
      removedMemberIds: removed,
    });
  }

  toDict(): Record<string, unknown> {
    return {
      previous_topology_id: this.previousTopologyId,
      previous_content_hash: this.previousContentHash,
      new_topology_id: this.newTopologyId,
      new_content_hash: this.newContentHash,
      added_members: this.addedMembers.map((m) => m.toDict()),
      removed_member_ids: [...this.removedMemberIds],
    };
  }
}

/**
 * Canonical JSON: deterministic key ordering (sorted), no whitespace.
 * Matches Python's `json.dumps(..., sort_keys=True, separators=(",",":"))`.
 *
 * Standard `JSON.stringify` doesn't sort keys; this helper does the
 * walk-and-sort manually.
 */
function canonicalJsonStringify(value: unknown): string {
  if (value === null) return "null";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) return "null";
    return String(value);
  }
  if (typeof value === "boolean") return value ? "true" : "false";
  if (typeof value === "string") return JSON.stringify(value);
  if (Array.isArray(value)) {
    return "[" + value.map(canonicalJsonStringify).join(",") + "]";
  }
  if (typeof value === "object") {
    const entries = Object.entries(value as Record<string, unknown>);
    entries.sort((a, b) => (a[0] < b[0] ? -1 : a[0] > b[0] ? 1 : 0));
    return (
      "{" +
      entries
        .map(([k, v]) => JSON.stringify(k) + ":" + canonicalJsonStringify(v))
        .join(",") +
      "}"
    );
  }
  // Fallback for undefined / symbol / function — match JSON.stringify
  // behavior (omit the key) by returning null; caller responsible for
  // not passing these in.
  return "null";
}
